src/iframe-helper.js

```diff
@@ -212,25 +251,26 @@ export function listenFor(iframe, typeOfMessage, callback, opt_is3P) {
 
   const listenForEvents = getOrCreateListenForEvents(
     parentWin,
-    iframe
+    iframe,
+    opt_is3P
   );
 
-  const sentinel = getSentinel_(opt_is3P);
+
   let events = listenForEvents[typeOfMessage] ||
     (listenForEvents[typeOfMessage] = []);
 
   let unlisten;
-  let listener = function(data) {
-    // If this iframe no longer has a contentWindow is was removed
-    // from the DOM. Unlisten immediately as we can never again receive
-    // messages for it (
+  let listener = function(data, source, origin) {
+    // Exclude nested frames if necessary.
+    if (!opt_includingNestedWindows && source != iframe.contentWindow) {
```

Comment:
It might be worth adding a comment here that at this point, source is either iframe.contentWindow, or one of its child frames, and cannot be an unrelated window.